### PR TITLE
Add session attendance routes and associated unit tests

### PR DESCRIPTION
**Pull request by aoby**
[Spec](https://docs.google.com/document/d/1w1Xgk1vyy148VjPIeM-SRP-Yi-oqYaEq8mDhw-uLFpE/edit#). [Task](https://trello.com/c/2UKGPhJS/12-update-attendance-model).

Navigating to the `code.org/pd/XXXX` attendance link when an enrollment is associated with your account attends immediately:
image

If there is no matching enrollment, I see a "safe" name selector:
image

Clicking a name registers me as that person, and links that enrollment with my user account so I won't see this again.

Clicking "My name is not on the list" brings me to a join workshop page:
image

And completing that joins me to the workshop. 

Finally, if I happen to have a student account, it will be upgraded to teacher automatically in the last step if the email I select matches my hashed account email. If it doesn't, I'll see one more page:
image

And this will only accept matching emails:
image

### CODE REVIEW

File: dashboard/app/controllers/pd/session_attendance_controller.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
+class Pd::SessionAttendanceController < ApplicationController
+  load_and_authorize_resource :session, class: 'Pd::Session', id_param: :session_code, find_by: :code
+  rescue_from ActiveRecord::RecordNotFound do
+    render :not_found, status: :not_found
+  end
```
Comment (Hamms) [inline dashboard/app/controllers/pd/session_attendance_controller.rb:5]: will this not happen by default?

File: dashboard/app/controllers/pd/workshop_enrollment_controller.rb [hunk 2]
```diff
@@ -157,8 +161,46 @@ def confirm_join
...
+  def join_session
+    @enrollment = get_workshop_user_enrollment
+  end
+
+  # POST /pd/attend/:session_code/join
+  # TODO (Andrew): once we're satisfied with this new session attendance model, remove #confirm_join_section
+  def confirm_join_session
+    @enrollment = build_enrollment_from_params
+
+    if !@enrollment.valid? || !@enrollment.save
```
Comment (Hamms) [inline dashboard/app/controllers/pd/workshop_enrollment_controller.rb]: why both?

File: dashboard/app/models/pd/enrollment.rb [hunk 3]
```diff
@@ -36,6 +37,7 @@ class Pd::Enrollment < ActiveRecord::Base
   belongs_to :user
   has_one :workshop_material_order, class_name: 'Pd::WorkshopMaterialOrder', foreign_key: :pd_enrollment_id
   has_many :attendances, class_name: 'Pd::Attendance', foreign_key: :pd_enrollment_id
+  auto_strip_attributes :first_name, :last_name
```
Comment (Hamms) [inline dashboard/app/models/pd/enrollment.rb:40]: :+1:

File: dashboard/app/views/pd/session_attendance/match_registration.html.haml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+- content_for(:head) do
+  = stylesheet_link_tag 'css/pd', media: 'all'
+
+:css
+  #application-container .form-inline {
+    display: inline-block;
+    padding: 0 2px;
+  }
```
Comment (Hamms) [inline dashboard/app/views/pd/session_attendance/match_registration.html.haml]: should this just go in the `css/pd` stylesheet?

File: dashboard/app/views/pd/session_attendance/upgrade_account.html.haml [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+- if @email_mismatch
+  #error_explanation
+    %h2 Error
+    %p
+      - if @email.blank?
+        Email is required
+      - else
+        Email must match your login. If you want to use this email instead, first update it in
+        = link_to 'account settings', "/users/edit?user_return_to: #{request.url}"
```
Comment (Hamms) [inline dashboard/app/views/pd/session_attendance/upgrade_account.html.haml]: as we discussed offline, I think this return param won't actually work just yet; have you made a note somewhere for either you or I to add that functionality?

File: dashboard/app/controllers/pd/session_attendance_controller.rb (hunk 1 above)
Comment (aoby) [reply dashboard/app/controllers/pd/session_attendance_controller.rb:5]: This is rendering a specific (friendly) not_found page specific to this controller, from [this view](https://github.com/code-dot-org/code-dot-org/pull/15170/files#diff-292fcca9c117109bda4fbdbde95b84ab). By default, we get the generic [404 page](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/public/404.html).

File: dashboard/app/controllers/pd/workshop_enrollment_controller.rb (hunk 2 above)
Comment (aoby) [reply dashboard/app/controllers/pd/workshop_enrollment_controller.rb]: Good question. I don't think I need the first check. Verifying...

Comment (aoby) [reply dashboard/app/controllers/pd/workshop_enrollment_controller.rb]: Yeah, I don't know why I did that! Removing the first check

File: dashboard/app/views/pd/session_attendance/upgrade_account.html.haml (hunk 5 above)
Comment (aoby) [reply dashboard/app/views/pd/session_attendance/upgrade_account.html.haml]: Oh yeah, thanks for reminding me :(

I'll remove the param and add a TODO

**Pull request merged**